File: ruby/050_StairJump.rb
# 050 - Stair Jump
# DP法

MOD = 10 ** 9 + 7
N, L = gets.split.map(&:to_i)

dp = [0] * (N + 1)
dp[0] = 1

(1..N).each do |i|
  dp[i] += dp[i - 1]
  dp[i] += dp[i - L] if i >= L
  dp[i] %= MOD
end

puts dp[N]

File: kotlin/050_StairJump.kt
/**
 * 050 - Stair Jump
 * DP法
 */

fun main() {
  val MOD = 1_000_000_000 + 7
  val (N, L) = readLine()!!.split(" ").map(String::toInt)

  val dp = MutableList<Int>(N + 1) { 0 }
  dp[0] = 1

  for (i in 0 .. N) {
    if (i + L <= N) dp[i + L] = (dp[i + L] + dp[i]) % MOD
    if (i + 1 <= N) dp[i + 1] = (dp[i + 1] + dp[i]) % MOD
  }

  println(dp[N])
}
